dotCMS/src/main/java/com/dotcms/rendering/velocity/servlet/VelocityPreviewServlet.java

```diff
@@ -0,0 +1,359 @@
+import com.dotmarketing.util.PageMode;
+import com.dotmarketing.util.UtilMethods;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.URLDecoder;
+import java.util.Map;
+
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.velocity.Template;
+import org.apache.velocity.context.Context;
+import org.apache.velocity.exception.MethodInvocationException;
+import org.apache.velocity.exception.ParseErrorException;
+import org.apache.velocity.exception.ResourceNotFoundException;
+
+import com.liferay.portal.language.LanguageException;
+import com.liferay.portal.language.LanguageUtil;
+import com.liferay.portal.model.User;
+
+public class VelocityPreviewServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+
+    private static PermissionAPI permissionAPI = APILocator.getPermissionAPI();
+
+
+
+    private static HostWebAPI hostWebAPI = WebAPILocator.getHostWebAPI();
+
+    private static VisitorAPI visitorAPI = APILocator.getVisitorAPI();
+
+    public static void setPermissionAPI(PermissionAPI permissionAPIRef) {
+        permissionAPI = permissionAPIRef;
+    }
+
+    private String CHARSET = "UTF-8";
+
+
+
+    public static final String VELOCITY_CONTEXT = "velocityContext";
+
+
+    @CloseDBIfOpened
+    protected void service(HttpServletRequest req, HttpServletResponse response) throws ServletException, IOException {
+
+        final String uri = URLDecoder.decode((req.getAttribute(Constants.CMS_FILTER_URI_OVERRIDE) != null)
+                ? (String) req.getAttribute(Constants.CMS_FILTER_URI_OVERRIDE)
+                : req.getRequestURI(), "UTF-8");
+
+
+
+        RequestWrapper request = new RequestWrapper(req);
+        request.setRequestUri(uri);
+
+
+        Host host = hostWebAPI.getCurrentHostNoThrow(request);
+
+        PageMode mode = PageMode.get(request);
+        // Checking if host is active
+        boolean hostlive;
+
+
+        try {
+            hostlive = APILocator.getVersionableAPI()
+                .hasLiveVersion(host);
+        } catch (Exception e1) {
+            UtilMethods.closeDbSilently();
+            throw new ServletException(e1);
+        }
+        if (!hostlive) {
+            try {
+                response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+                        LanguageUtil.get("server-unavailable-error-message"));
+            } catch (LanguageException e) {
+                Logger.error(CMSFilter.class, e.getMessage(), e);
+                response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
+            }
+            return;
+        }
+
+
+        if (DbConnectionFactory.isMsSql() && LicenseUtil.getLevel() < LicenseLevel.PROFESSIONAL.level) {
+            request.getRequestDispatcher("/portal/no_license.jsp")
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "/portal/no_license.jsp" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)